codec/src/main/java/io/netty/handler/codec/compression/BrotliEncoder.java

```diff
@@ -184,6 +194,28 @@ public void finish(ChannelHandlerContext ctx) throws IOException {
             writer.close();
             this.writer = null;
         }
+        return promise;
+    }
+
+    @Override
+    public void close(final ChannelHandlerContext ctx, final ChannelPromise promise) throws Exception {
+        ChannelFuture f = finishEncode(ctx, ctx.newPromise());
+        f.addListener(new ChannelFutureListener() {
+            @Override
+            public void operationComplete(ChannelFuture f) throws Exception {
+                ctx.close(promise);
+            }
+        });
+
+        if (!f.isDone()) {
+            // Ensure the channel is closed even if the write operation completes in time.
+            ctx.executor().schedule(new Runnable() {
```

Comment:
store the returned `Future` and cancel when handle is removed etc ?

Reply:
I just added a 'handlerRemoved' method.  Please take a look and let me know if I'm on the right track